File: test/java/brs/web/api/http/handler/DGSPurchaseTest.java
package brs.web.api.http.handler;

import brs.*;
import brs.DigitalGoodsStore.Goods;
import brs.common.QuickMocker;
import brs.common.QuickMocker.MockParam;
import brs.fluxcapacitor.FluxCapacitor;
import brs.fluxcapacitor.FluxValues;
import brs.services.AccountService;
import brs.services.ParameterService;
import brs.services.TimeService;
import brs.web.api.http.common.APITransactionManager;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import javax.servlet.http.HttpServletRequest;

import static brs.TransactionType.DigitalGoods.PURCHASE;
import static brs.web.api.http.common.JSONResponses.*;
import static brs.web.api.http.common.Parameters.*;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.powermock.api.mockito.PowerMockito.mockStatic;

@RunWith(PowerMockRunner.class)
@PrepareForTest(Signum.class)
public class DGSPurchaseTest extends AbstractTransactionTest {

  private DGSPurchase t;

  private ParameterService mockParameterService;
  private Blockchain mockBlockchain;
  private AccountService mockAccountService;
  private TimeService mockTimeService;
  private APITransactionManager apiTransactionManagerMock;

  @Before
  public void setUp() {
    mockParameterService = mock(ParameterService.class);
    mockBlockchain = mock(Blockchain.class);
    mockAccountService = mock(AccountService.class);
    mockTimeService = mock(TimeService.class);
    apiTransactionManagerMock = mock(APITransactionManager.class);

    t = new DGSPurchase(mockParameterService, mockBlockchain, mockAccountService, mockTimeService, apiTransactionManagerMock);
  }

  @Test
  public void processRequest() throws SignumException {
    final int goodsQuantity = 5;
    final long goodsPrice = 10L;
    final long deliveryDeadlineTimestamp = 100;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(QUANTITY_PARAMETER, goodsQuantity),
        new MockParam(PRICE_NQT_PARAMETER, goodsPrice),
        new MockParam(DELIVERY_DEADLINE_TIMESTAMP_PARAMETER, deliveryDeadlineTimestamp)
    );

    final long mockSellerId = 123L;
    final long mockGoodsId = 123L;
    final Goods mockGoods = mock(Goods.class);
    when(mockGoods.getId()).thenReturn(mockGoodsId);
    when(mockGoods.isDelisted()).thenReturn(false);
    when(mockGoods.getQuantity()).thenReturn(10);
    when(mockGoods.getPriceNQT()).thenReturn(10L);
    when(mockGoods.getSellerId()).thenReturn(mockSellerId);

    final Account mockSellerAccount = mock(Account.class);

    when(mockParameterService.getGoods(eq(req))).thenReturn(mockGoods);
    when(mockTimeService.getEpochTime()).thenReturn(10);

    when(mockAccountService.getAccount(eq(mockSellerId))).thenReturn(mockSellerAccount);

    mockStatic(Signum.class);
    final FluxCapacitor fluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.DIGITAL_GOODS_STORE);
    when(Signum.getFluxCapacitor()).thenReturn(fluxCapacitor);
    doReturn(Constants.FEE_QUANT_SIP3).when(fluxCapacitor).getValue(eq(FluxValues.FEE_QUANT));

    final Attachment.DigitalGoodsPurchase attachment = (Attachment.DigitalGoodsPurchase) attachmentCreatedTransaction(() -> t.processRequest(req), apiTransactionManagerMock);
    assertNotNull(attachment);

    assertEquals(PURCHASE, attachment.getTransactionType());
    assertEquals(goodsQuantity, attachment.getQuantity());
    assertEquals(goodsPrice, attachment.getPriceNqt());
    assertEquals(deliveryDeadlineTimestamp, attachment.getDeliveryDeadlineTimestamp());
    assertEquals(mockGoodsId, attachment.getGoodsId());
  }

  @Test
  public void processRequest_unknownGoods() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest();

    final Goods mockGoods = mock(Goods.class);
    when(mockGoods.isDelisted()).thenReturn(true);

    when(mockParameterService.getGoods(eq(req))).thenReturn(mockGoods);

    assertEquals(UNKNOWN_GOODS, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectPurchaseQuantity() throws SignumException {
    final int goodsQuantity = 5;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(QUANTITY_PARAMETER, goodsQuantity)
    );

    final Goods mockGoods = mock(Goods.class);
    when(mockGoods.isDelisted()).thenReturn(false);
    when(mockGoods.getQuantity()).thenReturn(4);

    when(mockParameterService.getGoods(eq(req))).thenReturn(mockGoods);

    assertEquals(INCORRECT_PURCHASE_QUANTITY, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectPurchasePrice() throws SignumException {
    final int goodsQuantity = 5;
    final long goodsPrice = 5L;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(QUANTITY_PARAMETER, goodsQuantity),
        new MockParam(PRICE_NQT_PARAMETER, goodsPrice)
    );

    final Goods mockGoods = mock(Goods.class);
    when(mockGoods.isDelisted()).thenReturn(false);
    when(mockGoods.getQuantity()).thenReturn(10);
    when(mockGoods.getPriceNQT()).thenReturn(10L);

    when(mockParameterService.getGoods(eq(req))).thenReturn(mockGoods);

    assertEquals(INCORRECT_PURCHASE_PRICE, t.processRequest(req));
  }


  @Test
  public void processRequest_missingDeliveryDeadlineTimestamp() throws SignumException {
    final int goodsQuantity = 5;
    final long goodsPrice = 10L;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(QUANTITY_PARAMETER, goodsQuantity),
        new MockParam(PRICE_NQT_PARAMETER, goodsPrice)
    );

    final Goods mockGoods = mock(Goods.class);
    when(mockGoods.isDelisted()).thenReturn(false);
    when(mockGoods.getQuantity()).thenReturn(10);
    when(mockGoods.getPriceNQT()).thenReturn(10L);

    when(mockParameterService.getGoods(eq(req))).thenReturn(mockGoods);

    assertEquals(MISSING_DELIVERY_DEADLINE_TIMESTAMP, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectDeliveryDeadlineTimestamp_unParsable() throws SignumException {
    final int goodsQuantity = 5;
    final long goodsPrice = 10L;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(QUANTITY_PARAMETER, goodsQuantity),
        new MockParam(PRICE_NQT_PARAMETER, goodsPrice),
        new MockParam(DELIVERY_DEADLINE_TIMESTAMP_PARAMETER, "unParsable")
    );

    final Goods mockGoods = mock(Goods.class);
    when(mockGoods.isDelisted()).thenReturn(false);
    when(mockGoods.getQuantity()).thenReturn(10);
    when(mockGoods.getPriceNQT()).thenReturn(10L);

    when(mockParameterService.getGoods(eq(req))).thenReturn(mockGoods);

    assertEquals(INCORRECT_DELIVERY_DEADLINE_TIMESTAMP, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectDeliveryDeadlineTimestamp_beforeCurrentTime() throws SignumException {
    final int goodsQuantity = 5;
    final long goodsPrice = 10L;
    final long deliveryDeadlineTimestamp = 100;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(QUANTITY_PARAMETER, goodsQuantity),
        new MockParam(PRICE_NQT_PARAMETER, goodsPrice),
        new MockParam(DELIVERY_DEADLINE_TIMESTAMP_PARAMETER, deliveryDeadlineTimestamp)
    );

    final Goods mockGoods = mock(Goods.class);
    when(mockGoods.isDelisted()).thenReturn(false);
    when(mockGoods.getQuantity()).thenReturn(10);
    when(mockGoods.getPriceNQT()).thenReturn(10L);

    when(mockParameterService.getGoods(eq(req))).thenReturn(mockGoods);
    when(mockTimeService.getEpochTime()).thenReturn(1000);

    assertEquals(INCORRECT_DELIVERY_DEADLINE_TIMESTAMP, t.processRequest(req));
  }
}

File: src/test/java/brs/http/DGSPurchaseTest.kt
package brs.http

import brs.Account
import brs.Attachment
import brs.Blockchain
import brs.DependencyProvider
import brs.DigitalGoodsStore.Goods
import brs.common.QuickMocker
import brs.common.QuickMocker.MockParam
import brs.fluxcapacitor.FluxValues
import brs.http.JSONResponses.INCORRECT_DELIVERY_DEADLINE_TIMESTAMP
import brs.http.JSONResponses.INCORRECT_PURCHASE_PRICE
import brs.http.JSONResponses.INCORRECT_PURCHASE_QUANTITY
import brs.http.JSONResponses.MISSING_DELIVERY_DEADLINE_TIMESTAMP
import brs.http.JSONResponses.UNKNOWN_GOODS
import brs.http.common.Parameters.DELIVERY_DEADLINE_TIMESTAMP_PARAMETER
import brs.http.common.Parameters.PRICE_NQT_PARAMETER
import brs.http.common.Parameters.QUANTITY_PARAMETER
import brs.services.AccountService
import brs.services.ParameterService
import brs.services.TimeService
import brs.transaction.TransactionType
import brs.transaction.digitalGoods.DigitalGoodsPurchase
import com.nhaarman.mockitokotlin2.*
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import javax.servlet.http.HttpServletRequest

class DGSPurchaseTest : AbstractTransactionTest() {

    private lateinit var t: DGSPurchase
    private lateinit var dp: DependencyProvider
    private lateinit var mockParameterService: ParameterService
    private lateinit var mockBlockchain: Blockchain
    private lateinit var mockAccountService: AccountService
    private lateinit var mockTimeService: TimeService
    private lateinit var apiTransactionManagerMock: APITransactionManager

    @Before
    fun setUp() {
        mockParameterService = mock<ParameterService>()
        whenever(mockParameterService!!.getSenderAccount(any())).doReturn(mock())
        mockBlockchain = mock<Blockchain>()
        mockAccountService = mock<AccountService>()
        mockTimeService = mock<TimeService>()
        apiTransactionManagerMock = mock<APITransactionManager>()
        dp = QuickMocker.dependencyProvider(mockParameterService!!, mockBlockchain!!, mockAccountService!!, mockTimeService!!, apiTransactionManagerMock!!)
        t = DGSPurchase(dp)
    }

    @Test
    fun processRequest() {
        val goodsQuantity = 5
        val goodsPrice = 10L
        val deliveryDeadlineTimestamp: Long = 100

        val request = QuickMocker.httpServletRequest(
                MockParam(QUANTITY_PARAMETER, goodsQuantity),
                MockParam(PRICE_NQT_PARAMETER, goodsPrice),
                MockParam(DELIVERY_DEADLINE_TIMESTAMP_PARAMETER, deliveryDeadlineTimestamp)
        )

        val mockSellerId = 123L
        val mockGoodsId = 123L
        val mockGoods = mock<Goods>()
        whenever(mockGoods.id).doReturn(mockGoodsId)
        whenever(mockGoods.isDelisted).doReturn(false)
        whenever(mockGoods.quantity).doReturn(10)
        whenever(mockGoods.priceNQT).doReturn(10L)
        whenever(mockGoods.sellerId).doReturn(mockSellerId)

        val mockSellerAccount = mock<Account>()

        whenever(mockParameterService!!.getGoods(eq<HttpServletRequest>(request))).doReturn(mockGoods)
        whenever(mockTimeService!!.epochTime).doReturn(10)

        whenever(mockAccountService!!.getAccount(eq(mockSellerId))).doReturn(mockSellerAccount)
        dp.fluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.DIGITAL_GOODS_STORE)
        dp.transactionTypes = TransactionType.getTransactionTypes(dp)

        val attachment = attachmentCreatedTransaction({ t!!.processRequest(request) }, apiTransactionManagerMock!!) as Attachment.DigitalGoodsPurchase
        assertNotNull(attachment)

        assertTrue(attachment.transactionType is DigitalGoodsPurchase)
        assertEquals(goodsQuantity.toLong(), attachment.quantity.toLong())
        assertEquals(goodsPrice, attachment.priceNQT)
        assertEquals(deliveryDeadlineTimestamp, attachment.deliveryDeadlineTimestamp.toLong())
        assertEquals(mockGoodsId, attachment.goodsId)
    }

    @Test
    fun processRequest_unknownGoods() {
        val request = QuickMocker.httpServletRequest()

        val mockGoods = mock<Goods>()
        whenever(mockGoods.isDelisted).doReturn(true)

        whenever(mockParameterService!!.getGoods(eq<HttpServletRequest>(request))).doReturn(mockGoods)

        assertEquals(UNKNOWN_GOODS, t!!.processRequest(request))
    }

    @Test
    fun processRequest_incorrectPurchaseQuantity() {
        val goodsQuantity = 5

        val request = QuickMocker.httpServletRequest(
                MockParam(QUANTITY_PARAMETER, goodsQuantity)
        )

        val mockGoods = mock<Goods>()
        whenever(mockGoods.isDelisted).doReturn(false)
        whenever(mockGoods.quantity).doReturn(4)

        whenever(mockParameterService!!.getGoods(eq<HttpServletRequest>(request))).doReturn(mockGoods)

        assertEquals(INCORRECT_PURCHASE_QUANTITY, t!!.processRequest(request))
    }

    @Test
    fun processRequest_incorrectPurchasePrice() {
        val goodsQuantity = 5
        val goodsPrice = 5L

        val request = QuickMocker.httpServletRequest(
                MockParam(QUANTITY_PARAMETER, goodsQuantity),
                MockParam(PRICE_NQT_PARAMETER, goodsPrice)
        )

        val mockGoods = mock<Goods>()
        whenever(mockGoods.isDelisted).doReturn(false)
        whenever(mockGoods.quantity).doReturn(10)
        whenever(mockGoods.priceNQT).doReturn(10L)

        whenever(mockParameterService!!.getGoods(eq<HttpServletRequest>(request))).doReturn(mockGoods)

        assertEquals(INCORRECT_PURCHASE_PRICE, t!!.processRequest(request))
    }


    @Test
    fun processRequest_missingDeliveryDeadlineTimestamp() {
        val goodsQuantity = 5
        val goodsPrice = 10L

        val request = QuickMocker.httpServletRequest(
                MockParam(QUANTITY_PARAMETER, goodsQuantity),
                MockParam(PRICE_NQT_PARAMETER, goodsPrice)
        )

        val mockGoods = mock<Goods>()
        whenever(mockGoods.isDelisted).doReturn(false)
        whenever(mockGoods.quantity).doReturn(10)
        whenever(mockGoods.priceNQT).doReturn(10L)

        whenever(mockParameterService!!.getGoods(eq<HttpServletRequest>(request))).doReturn(mockGoods)

        assertEquals(MISSING_DELIVERY_DEADLINE_TIMESTAMP, t!!.processRequest(request))
    }

    @Test
    fun processRequest_incorrectDeliveryDeadlineTimestamp_unParsable() {
        val goodsQuantity = 5
        val goodsPrice = 10L

        val request = QuickMocker.httpServletRequest(
                MockParam(QUANTITY_PARAMETER, goodsQuantity),
                MockParam(PRICE_NQT_PARAMETER, goodsPrice),
                MockParam(DELIVERY_DEADLINE_TIMESTAMP_PARAMETER, "unParsable")
        )

        val mockGoods = mock<Goods>()
        whenever(mockGoods.isDelisted).doReturn(false)
        whenever(mockGoods.quantity).doReturn(10)
        whenever(mockGoods.priceNQT).doReturn(10L)

        whenever(mockParameterService!!.getGoods(eq<HttpServletRequest>(request))).doReturn(mockGoods)

        assertEquals(INCORRECT_DELIVERY_DEADLINE_TIMESTAMP, t!!.processRequest(request))
    }

    @Test
    fun processRequest_incorrectDeliveryDeadlineTimestamp_beforeCurrentTime() {
        val goodsQuantity = 5
        val goodsPrice = 10L
        val deliveryDeadlineTimestamp: Long = 100

        val request = QuickMocker.httpServletRequest(
                MockParam(QUANTITY_PARAMETER, goodsQuantity),
                MockParam(PRICE_NQT_PARAMETER, goodsPrice),
                MockParam(DELIVERY_DEADLINE_TIMESTAMP_PARAMETER, deliveryDeadlineTimestamp)
        )

        val mockGoods = mock<Goods>()
        whenever(mockGoods.isDelisted).doReturn(false)
        whenever(mockGoods.quantity).doReturn(10)
        whenever(mockGoods.priceNQT).doReturn(10L)

        whenever(mockParameterService!!.getGoods(eq<HttpServletRequest>(request))).doReturn(mockGoods)
        whenever(mockTimeService!!.epochTime).doReturn(1000)

        assertEquals(INCORRECT_DELIVERY_DEADLINE_TIMESTAMP, t!!.processRequest(request))
    }
}
